Lib/shelve.py

```diff
@@ -186,7 +199,16 @@ class BsdDbShelf(Shelf):
     """
 
     def __init__(self, dict, protocol=None, writeback=False,
-                 keyencoding="utf-8"):
+                 keyencoding="utf-8", *, serializer=None, deserializer=None):
+        if serializer is None and deserializer is None:
+            self.serializer = dumps
+            self.deserializer = loads
+        elif (serializer is None) ^ (deserializer is None):
+            raise ShelveError("Serializer and deserializer must be"
+                              "defined together.")
+        else:
+            self.serializer = serializer
+            self.deserializer = deserializer
```

Comment:
These should be passed to `Shelf.__init__`, rather than set here. If they're not passed in, `Shelf.__init__` will set them to the defaults.

The subclasses should also be tested.